### ISSUE TITLE
Add option(s) to override provable relationships

### ISSUE THREAD

**Issue by fubuloubu** [ConsenSys/mythril-classic#172]
In my contract I have a relationship where each user is storing their proportion of a total quantity. Due to a provable mathematical relationship between the quantities (I increase total every time I increase mine, and I decrease total every time I decrease mine), I can maintain an invariant in my code with more minimal input validation.

How would I add comments to my code to request mythril ignore specific violations that I could provable demonstrate are invariant?

Example contract:

```solidity
contract A {
  uint total;
  mapping (address => uint) mine;

  function add()
    public
  {
    require( total + 1 > total ); // Overflow protection
    // require( mine[msg.sender] <= total ); // This should be necessary to avoid integer overflow
    total += 1;
    mine[msg.sender] += 1; // Curiously, no integer overflow raised here...
  }

  function subtract()
    public
  {
    require( mine[msg.sender] > 0 ); // Underflow protection
    // require( mine[msg.sender] <= total ); // This avoids throwing integer underflow
    total -= 1; // mythril throws an integer underflow here
    mine[msg.sender] -= 1; // @myth-ignore UNDERFLOW - Example of syntax
  }
}
```

Some tools have a `@ignore CHECK_NAME` comment that can be added to a line to ignore a specific check. I could save a bit of gas by just being able to avoid having a `assert(mine <= total)` in my code in two spots

**Comment by b-mueller** [comment]
Hi @fubuloubu, I love the idea with the `@myth-ignore` comment.

Also curious why the overflows in `add()` are not detected, we'll look into it.

**Comment by b-mueller** [comment]
My first guess is that there's an off-by-one error in the expression checking for overflows.

When I change line 11 to:

```
    mine[msg.sender] += 2;
```

Then the overflow is reported.

**Comment by b-mueller** [comment]
@JoranHonig I have narrowed the problem of `+1` not being caught down to the`_verify_integer_overflow` function, specifically the second part of false positive detection. Still not sure why `+1` would be a special case.

**Comment by b-mueller** [comment]
We'll definitely implement that. Currently Mythril is still largely experimental, the focus is now to create tools that are actually useful for auditors and developers. So keep the feedback coming!

**Comment by JoranHonig** [comment]
@b-mueller Maybe we should create a separate issue for the missed overflow.

I think implementing tags like ```@myth-ignore``` is a great idea!

**Comment by b-mueller** [comment]
> @b-mueller Maybe we should create a separate issue for the missed overflow.

See #176

**Comment by maurelian** [comment]
This feature will be super helpful (essential IMO) for IDE integration.

**Comment by rocky** [comment]
A very good idea, I think. I recently noticed in one of Suhabe's benchmarks [this](https://github.com/ConsenSys/evm-analyzer-benchmark-suite/blob/master/benchmarks/integer_overflow_storageinvariant.sol#L1): 

     //No underflow because of credit/balance invariant: sum (credit.values) = balance

How do you feel about this variation: in a front-end wrapper, look for comments of the form: 

    //invariant: sum (credit.values) = balance

And add that as a constraint to the z3 solver. Presumably this would previent Mythril from triggering the violation. 

Outside of a wrapper an a `--constraint` option could be supplied.

**Comment by fubuloubu** [comment]
I like the "in source comments" approach best, because it will end up in version control and a part of the build release instead of some external documentation or script for execution. At least as an option to do either is good.

That syntax isn't bad for invariants. To summarize, this invariant would be treated as an *additional* constraint to the solver? E.g. not ignoring a current issue, but hinting to the solver that some false positive condition is actually not possible due to the constraint. This would err on the side of caution, which I like, but may not be possible for all false positive cases.

I was hoping to have a method that works in general whenever an issue pops up that is a false positive but is difficult to constrain out.

Note: I think it should be `//@invariant: sum(credit.values) == balance`
The `@invariant` syntax (alternative `#invariant`) makes it easier to identify and key off of, and it should be equality not assignment that is being checked (that's probably just a typo).

P.S. Does the Mythril engine hold mapping assignments in parallel to execution? Is there a limit to the size of a mapping that can be held? Is this configurable?

**Comment by rocky** [comment]
Since the above is an off-shoot of this issue, I have created a separate issue to track discussion there. See #405 .

**Comment by JoranHonig** [comment]
Adding invariants and statements to be proved is discussed in #405 

Adding constraints to inform Mythril of possible invariants in order to prune & resolve false positives is not necessary anymore as Mythril now has a constrained memory model (vs the old unconstrained one)

Therefore I'm closing this issue if there is anything that isn't captured in the other issue yet. Please open another issue so we can look at that separately
